File: src/services/probe_task.rs
use sqlx::{PgPool, Row};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use tokio::sync::Mutex;

use crate::services::metadata::probe_file;
use crate::services::scanner::resolve_strm_path;

#[derive(Debug, Clone, serde::Serialize)]
pub struct ProbeProgress {
    pub status: String,
    pub total_items: i64,
    pub processed_items: i64,
    pub success_items: i64,
    pub failed_items: i64,
    pub current_item: Option<String>,
    pub percentage: i32,
    pub threads: i32,
    pub error: Option<String>,
}

#[derive(Clone)]
pub struct ProbeTask {
    progress: Arc<Mutex<ProbeProgress>>,
    stop_flag: Arc<AtomicBool>,
}

impl ProbeTask {
    pub fn new() -> Self {
        Self {
            progress: Arc::new(Mutex::new(ProbeProgress {
                status: "idle".into(),
                total_items: 0,
                processed_items: 0,
                success_items: 0,
                failed_items: 0,
                current_item: None,
                percentage: 0,
                threads: 5,
                error: None,
            })),
            stop_flag: Arc::new(AtomicBool::new(false)),
        }
    }

    pub async fn get_progress(&self) -> ProbeProgress {
        self.progress.lock().await.clone()
    }

    pub async fn stop(&self) {
        let mut p = self.progress.lock().await;
        if p.status == "running" {
            self.stop_flag.store(true, Ordering::SeqCst);
            p.status = "stopping".into();
            tracing::info!("[Probe] Stop requested");
        }
    }

    pub async fn start(&self, pool: PgPool, threads: i32) -> Result<(), String> {
        {
            let p = self.progress.lock().await;
            if p.status == "running" || p.status == "stopping" {
                return Err("Probe task is already running".into());
            }
        }

        let threads = threads.max(1).min(20);
        self.stop_flag.store(false, Ordering::SeqCst);

        // Get items missing mediainfo
        let rows: Vec<(uuid::Uuid, uuid::Uuid, String, String)> = sqlx::query_as(
            "SELECT mv.id, mv.item_id, mv.file_path, mv.name FROM media_versions mv WHERE mv.mediainfo IS NULL ORDER BY mv.id",
        )
        .fetch_all(&pool)
        .await
        .map_err(|e| e.to_string())?;

        if rows.is_empty() {
            let mut p = self.progress.lock().await;
            *p = ProbeProgress {
                status: "completed".into(),
                total_items: 0, processed_items: 0, success_items: 0, failed_items: 0,
                current_item: None, percentage: 100, threads, error: None,
            };
            return Ok(());
        }

        // Get path mappings
        let mappings = get_probe_path_mappings(&pool).await;

        {
            let mut p = self.progress.lock().await;
            *p = ProbeProgress {
                status: "running".into(),
                total_items: rows.len() as i64,
                processed_items: 0, success_items: 0, failed_items: 0,
                current_item: None, percentage: 0, threads, error: None,
            };
        }

        tracing::info!("[Probe] Starting: {} items, {} threads", rows.len(), threads);

        let progress = self.progress.clone();
        let stop_flag = self.stop_flag.clone();

        tokio::spawn(async move {
            let sem = Arc::new(tokio::sync::Semaphore::new(threads as usize));
            let mut handles = Vec::new();

            for (mv_id, item_id, file_path, name) in rows {
                if stop_flag.load(Ordering::SeqCst) {
                    break;
                }

                let permit = sem.clone().acquire_owned().await.unwrap();
                let pool = pool.clone();
                let mappings = mappings.clone();
                let progress = progress.clone();
                let stop_flag = stop_flag.clone();

                handles.push(tokio::spawn(async move {
                    let _permit = permit;
                    if stop_flag.load(Ordering::SeqCst) {
                        return;
                    }

                    let result = probe_one_item(&pool, &mv_id.to_string(), &item_id.to_string(), &file_path, &name, &mappings).await;

                    let mut p = progress.lock().await;
                    p.processed_items += 1;
                    match result {
                        Ok(_) => p.success_items += 1,
                        Err(_) => p.failed_items += 1,
                    }
                    p.percentage = ((p.processed_items as f64 / p.total_items as f64) * 100.0) as i32;
                    p.current_item = Some(name);
                }));
            }

            for h in handles {
                let _ = h.await;
            }

            let mut p = progress.lock().await;
            if stop_flag.load(Ordering::SeqCst) {
                p.status = "idle".into();
                tracing::info!("[Probe] Stopped. {}/{} processed", p.processed_items, p.total_items);
            } else {
                p.status = "completed".into();
                tracing::info!("[Probe] Completed. {} success, {} failed", p.success_items, p.failed_items);
            }
        });

        Ok(())
    }
}

async fn get_probe_path_mappings(pool: &PgPool) -> Vec<(String, String)> {
    let val: Option<String> = sqlx::query_scalar("SELECT value FROM system_config WHERE key = 'probe_path_mappings'")
        .fetch_optional(pool)
        .await
        .ok()
        .flatten();

    if let Some(ref v) = val {
        if let Ok(arr) = serde_json::from_str::<Vec<serde_json::Value>>(v) {
            return arr
                .iter()
                .filter_map(|m| {
                    let from = m.get("from")?.as_str()?.to_string();
                    let to = m.get("to")?.as_str()?.to_string();
                    Some((from, to))
                })
                .collect();
        }
    }
    vec![]
}

fn apply_path_mappings(path: &str, mappings: &[(String, String)]) -> String {
    for (from, to) in mappings {
        if path.starts_with(from) {
            return format!("{to}{}", &path[from.len()..]);
        }
    }
    path.to_string()
}

async fn probe_one_item(
    pool: &PgPool,
    mv_id: &str,
    item_id: &str,
    file_path: &str,
    name: &str,
    mappings: &[(String, String)],
) -> Result<(), String> {
    let real_path = if file_path.ends_with(".strm") {
        resolve_strm_path(file_path).ok_or_else(|| format!("Cannot resolve strm: {file_path}"))?
    } else {
        file_path.to_string()
    };

    let real_path = apply_path_mappings(&real_path, mappings);

    if !std::path::Path::new(&real_path).exists() {
        return Err(format!("File not found: {real_path}"));
    }

    let result = tokio::time::timeout(
        std::time::Duration::from_secs(30),
        probe_file(&real_path),
    )
    .await
    .map_err(|_| "Probe timeout".to_string())?
    .map_err(|e| e.to_string())?;

    // Build simplified mediainfo for DB
    let streams: Vec<serde_json::Value> = result.streams.iter().map(|s| {
        serde_json::json!({
            "Codec": s.codec,
            "Type": s.stream_type,
            "Index": s.index,
            "IsDefault": s.is_default,
            "IsForced": s.is_forced,
            "Width": s.width,
            "Height": s.height,
            "BitRate": s.bit_rate,
            "Channels": s.channels,
            "SampleRate": s.sample_rate,
            "Language": s.language,
            "Title": s.title,
            "DisplayTitle": s.display_title,
        })
    }).collect();

    let file_size: Option<i64> = std::fs::metadata(&real_path).ok().map(|m| m.len() as i64);
    let bitrate: Option<i64> = if let (Some(size), true) = (file_size, result.duration_ticks > 0) {
        let dur_sec = result.duration_ticks as f64 / 10_000_000.0;
        Some((size as f64 * 8.0 / dur_sec) as i64)
    } else {
        None
    };

    let db_info = serde_json::json!({
        "Name": name,
        "Size": file_size,
        "RunTimeTicks": result.duration_ticks,
        "Bitrate": bitrate,
        "Container": result.container,
        "MediaStreams": streams,
    });

    sqlx::query("UPDATE media_versions SET mediainfo = $1, runtime_ticks = $2, bitrate = $3, size = $4 WHERE id = $5::uuid")
        .bind(serde_json::to_string(&db_info).unwrap_or_default())
        .bind(result.duration_ticks)
        .bind(bitrate)
        .bind(file_size)
        .bind(mv_id)
        .execute(pool)
        .await
        .map_err(|e| e.to_string())?;

    // Update item runtime if not set
    sqlx::query("UPDATE items SET runtime_ticks = $1, updated_at = NOW() WHERE id = $2 AND (runtime_ticks IS NULL OR runtime_ticks = 0)")
        .bind(result.duration_ticks)
        .bind(item_id)
        .execute(pool)
        .await
        .map_err(|e| e.to_string())?;

    Ok(())
}

pub async fn get_missing_mediainfo_count(pool: &PgPool) -> Result<i64, sqlx::Error> {
    let count: (i64,) = sqlx::query_as("SELECT count(*) FROM media_versions WHERE mediainfo IS NULL")
        .fetch_one(pool)
        .await?;
    Ok(count.0)
}

File: internal/services/probe_task.go
package services

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"

	"fyms/internal/repository"
)

type ProbeProgress struct {
	Status         string  `json:"status"`
	TotalItems     int64   `json:"totalItems"`
	ProcessedItems int64   `json:"processedItems"`
	SuccessItems   int64   `json:"successItems"`
	FailedItems    int64   `json:"failedItems"`
	CurrentItem    *string `json:"currentItem,omitempty"`
	Percentage     int     `json:"percentage"`
	Threads        int     `json:"threads"`
	MissingCount   int64   `json:"missingCount"`
	VersionsTotal  int64   `json:"versionsTotal"`
	Error          *string `json:"error,omitempty"`
}

type ProbeTask struct {
	mu       sync.Mutex
	progress ProbeProgress
	stopFlag atomic.Bool
}

func NewProbeTask() *ProbeTask {
	return &ProbeTask{
		progress: ProbeProgress{Status: "idle", Threads: 5},
	}
}

func (pt *ProbeTask) GetProgress() ProbeProgress {
	pt.mu.Lock()
	defer pt.mu.Unlock()
	return pt.progress
}

func (pt *ProbeTask) Stop() {
	pt.mu.Lock()
	defer pt.mu.Unlock()
	if pt.progress.Status == "running" {
		pt.stopFlag.Store(true)
		pt.progress.Status = "stopping"
		slog.Info("[Probe] Stop requested")
	}
}

func (pt *ProbeTask) Start(pool *pgxpool.Pool, threads int) error {
	pt.mu.Lock()
	if pt.progress.Status == "running" || pt.progress.Status == "stopping" {
		pt.mu.Unlock()
		return fmt.Errorf("probe task is already running")
	}
	pt.mu.Unlock()

	if threads < 1 {
		threads = 1
	}
	if threads > 20 {
		threads = 20
	}
	pt.stopFlag.Store(false)

	ctx := context.Background()
	repo := repository.NewBackgroundTaskRepository(pool)
	items, err := repo.ListProbeTargets(ctx)
	if err != nil {
		return err
	}

	if len(items) == 0 {
		pt.mu.Lock()
		pt.progress = ProbeProgress{Status: "completed", Percentage: 100, Threads: int(threads)}
		pt.mu.Unlock()
		return nil
	}

	mappings := getProbePathMappings(ctx, pool)

	pt.mu.Lock()
	pt.progress = ProbeProgress{
		Status:     "running",
		TotalItems: int64(len(items)),
		Threads:    threads,
	}
	pt.mu.Unlock()

	slog.Info("[Probe] Starting", "items", len(items), "threads", threads)

	go func() {
		sem := make(chan struct{}, threads)
		var wg sync.WaitGroup
		var processed, success, failed atomic.Int64

		for _, item := range items {
			if pt.stopFlag.Load() {
				break
			}

			sem <- struct{}{}
			wg.Add(1)

			go func(pi repository.ProbeTarget) {
				defer func() { <-sem; wg.Done() }()
				if pt.stopFlag.Load() {
					return
				}

				err := probeOneItem(ctx, pool, pi.MediaVersionID.String(), pi.ItemID.String(), pi.FilePath, pi.Name, mappings)
				p := processed.Add(1)
				if err != nil {
					failed.Add(1)
				} else {
					success.Add(1)
				}

				pt.mu.Lock()
				pt.progress.ProcessedItems = p
				pt.progress.SuccessItems = success.Load()
				pt.progress.FailedItems = failed.Load()
				pt.progress.CurrentItem = &pi.Name
				if pt.progress.TotalItems > 0 {
					pt.progress.Percentage = int(float64(p) / float64(pt.progress.TotalItems) * 100)
				}
				pt.mu.Unlock()
			}(item)
		}

		wg.Wait()

		pt.mu.Lock()
		if pt.stopFlag.Load() {
			pt.progress.Status = "idle"
			slog.Info("[Probe] Stopped", "processed", pt.progress.ProcessedItems, "total", pt.progress.TotalItems)
		} else {
			pt.progress.Status = "completed"
			slog.Info("[Probe] Completed", "success", pt.progress.SuccessItems, "failed", pt.progress.FailedItems)
		}
		pt.mu.Unlock()
	}()

	return nil
}

func getProbePathMappings(ctx context.Context, pool *pgxpool.Pool) [][2]string {
	val, ok, err := repository.NewSystemConfigRepository(pool).GetString(ctx, "probe_path_mappings")
	if err != nil || !ok || val == "" {
		return nil
	}

	var arr []map[string]string
	if err := json.Unmarshal([]byte(val), &arr); err != nil {
		return nil
	}

	var mappings [][2]string
	for _, m := range arr {
		from, ok1 := m["from"]
		to, ok2 := m["to"]
		if ok1 && ok2 {
			mappings = append(mappings, [2]string{from, to})
		}
	}
	return mappings
}

func applyPathMappings(path string, mappings [][2]string) string {
	for _, m := range mappings {
		if len(path) >= len(m[0]) && path[:len(m[0])] == m[0] {
			return m[1] + path[len(m[0]):]
		}
	}
	return path
}

func probeOneItem(ctx context.Context, pool *pgxpool.Pool, mvID, itemID, filePath, name string, mappings [][2]string) error {
	realPath := filePath
	if filepath.Ext(filePath) == ".strm" {
		resolved := ResolveStrmPath(filePath)
		if resolved == nil {
			return fmt.Errorf("cannot resolve strm: %s", filePath)
		}
		realPath = *resolved
	}

	remote := strings.HasPrefix(realPath, "http://") || strings.HasPrefix(realPath, "https://")

	// 本地文件:应用路径映射并校验存在;远程直链:直接交给 ffprobe(支持 http)。
	var fileSize *int64
	if !remote {
		realPath = applyPathMappings(realPath, mappings)
		fi, err := os.Stat(realPath)
		if err != nil {
			return fmt.Errorf("file not found: %s", realPath)
		}
		s := fi.Size()
		fileSize = &s
	}

	doneCh := make(chan *ProbeResult, 1)
	errCh := make(chan error, 1)
	go func() {
		var (
			result *ProbeResult
			err    error
		)
		if remote {
			result, err = ProbeRemote(realPath)
		} else {
			result, err = ProbeFile(realPath)
		}
		if err != nil {
			errCh <- err
		} else {
			doneCh <- result
		}
	}()

	var result *ProbeResult
	select {
	case result = <-doneCh:
	case err := <-errCh:
		return err
	case <-time.After(40 * time.Second):
		return fmt.Errorf("probe timeout")
	}

	var streams []map[string]interface{}
	for _, s := range result.Streams {
		streams = append(streams, map[string]interface{}{
			"Codec": s.Codec, "Type": s.StreamType, "Index": s.Index,
			"IsDefault": s.IsDefault, "IsForced": s.IsForced,
			"Width": s.Width, "Height": s.Height, "BitRate": s.BitRate,
			"Channels": s.Channels, "SampleRate": s.SampleRate,
			"Language": s.Language, "Title": s.Title, "DisplayTitle": s.DisplayTitle,
		})
	}

	// 远程无法 os.Stat,size 留空;有大小且有时长才算码率。
	var bitrate *int64
	if fileSize != nil && result.DurationTicks > 0 {
		durSec := float64(result.DurationTicks) / 10_000_000.0
		b := int64(float64(*fileSize) * 8.0 / durSec)
		bitrate = &b
	}

	dbInfo := map[string]interface{}{
		"Name": name, "Size": fileSize, "RunTimeTicks": result.DurationTicks,
		"Bitrate": bitrate, "Container": result.Container, "MediaStreams": streams,
	}
	dbInfoJSON, _ := json.Marshal(dbInfo)

	chaptersSlice := result.Chapters
	if chaptersSlice == nil {
		chaptersSlice = []ProbeChapter{}
	}
	chaptersJSON, _ := json.Marshal(chaptersSlice)

	// 用 COALESCE 保护 size/bitrate:远程探测传 nil 时不覆盖已有列值。
	// chapters 每次探测都更新（空数组表示"探过但无章节"，区分 NULL "未探"）。
	repo := repository.NewBackgroundTaskRepository(pool)
	if err := repo.UpdateProbeMediaVersion(ctx, mvID, string(dbInfoJSON), result.DurationTicks, bitrate, fileSize, string(chaptersJSON)); err != nil {
		return err
	}

	return repo.FillItemRuntimeTicksIfEmpty(ctx, itemID, result.DurationTicks)
}

func GetMissingMediainfoCount(ctx context.Context, pool *pgxpool.Pool) (int64, error) {
	return repository.NewBackgroundTaskRepository(pool).GetMissingMediainfoCount(ctx)
}

func GetTotalMediaVersionsCount(ctx context.Context, pool *pgxpool.Pool) (int64, error) {
	return repository.NewBackgroundTaskRepository(pool).GetTotalMediaVersionsCount(ctx)
}
